quantization/src/encoded_vectors_binary.rs

```diff
@@ -123,29 +129,53 @@ impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
         let xor_product = Self::xor_product(v1, v2);
 
         // Dot product in a range [-1; 1] is approximated by NXOR in a range [0; 1]
+        // L1 distance in range [-1; 1] (alpha=2) is approximated by alpha*XOR in a range [0; 1]
+        // L2 distance in range [-1; 1] (alpha=2) is approximated by alpha*sqrt(XOR) in a range [0; 1]
+        // L2 distance is implemented as L2^2 in u8 quantization, so this implementation matches that
         // For example:
 
-        // A    |   B   | Dot product
-        // -0.5 | -0.5  |  0.25
-        // -0.5 |  0.5  | -0.25
-        //  0.5 | -0.5  | -0.25
-        //  0.5 |  0.5  |  0.25
+        // A    |   B   | Dot product | L1 | L2 |
+        // -0.5 | -0.5  |  0.25       | 0  | 0  |
+        // -0.5 |  0.5  | -0.25       | 1  | 1  |
+        //  0.5 | -0.5  | -0.25       | 1  | 1  |
+        //  0.5 |  0.5  |  0.25       | 0  | 0  |
 
-        // A | B  |  NXOR
-        // 0 | 0  |  1
-        // 0 | 1  |  0
-        // 1 | 0  |  0
-        // 1 | 1  |  1
+        // A | B  |  NXOR | XOR
+        // 0 | 0  |  1    | 0
+        // 0 | 1  |  0    | 1
+        // 1 | 0  |  0    | 1
+        // 1 | 1  |  1    | 0
 
         // So is `invert` is true, we return XOR, otherwise we return (dim - XOR)
 
+        let alpha = self.metadata.alpha.unwrap_or(1.0);
+
         let zeros_count = self.metadata.vector_parameters.dim - xor_product;
         if self.metadata.vector_parameters.invert {
-            xor_product as f32 - zeros_count as f32
+            match self.metadata.vector_parameters.distance_type {
+                DistanceType::Dot => xor_product as f32 - zeros_count as f32,
+                DistanceType::L1 => alpha * xor_product as f32,
+                DistanceType::L2 => alpha * alpha * xor_product as f32,
+            }
         } else {
-            zeros_count as f32 - xor_product as f32
+            match self.metadata.vector_parameters.distance_type {
+                DistanceType::Dot => zeros_count as f32 - xor_product as f32,
+                DistanceType::L1 => 1.0 / (alpha * xor_product as f32),
+                DistanceType::L2 => 1.0 / (alpha * alpha * xor_product as f32),
```

Comment:
What if the `/ rhs` ever becomes `0.0`?

Reply:
Yes, I also caught that, I have changed this locally actually to `dim*exp(-x)` so it ranges between dim -> 0.

I also dropped the whole alpha estimation as Dot already assumes [-1, 1] range and I don't want to change it. Decided to hardcode alpha=2 for L1 and L2.